packages/x-charts/src/hooks/useTicksGrouped.ts

```diff
@@ -0,0 +1,215 @@
+    valueFormatter?: AxisConfig['valueFormatter'];
+    direction: 'x' | 'y';
+    getGrouping: (value: any, dataIndex: number) => any[];
+  } & Pick<TickParams, 'tickNumber' | 'tickInterval' | 'tickPlacement' | 'tickLabelPlacement'>,
+): GroupedTickItemType[] {
+  const {
+    scale,
+    tickInterval,
+    tickLabelPlacement = 'middle',
+    tickPlacement = 'extremities',
+    getGrouping,
+  } = options;
+
+  return React.useMemo(() => {
+    const domain = scale.domain();
+    const filteredDomain =
+      (typeof tickInterval === 'function' && domain.filter(tickInterval)) ||
+      (typeof tickInterval === 'object' && tickInterval) ||
+      domain;
+
+    if (scale.bandwidth() > 0) {
+      // scale type = 'band'
+      const { entries, maxGroupIndex } = mapToGrouping(
+        filteredDomain,
+        getGrouping,
+        tickPlacement,
+        tickLabelPlacement,
+        scale,
+      );
+
+      if (entries[0]) {
+        entries[0].ignoreTick = true;
+      }
+
+      return [
+        {
+          formattedValue: undefined,
+          offset: scale.range()[0],
+          labelOffset: 0,
+          groupIndex: maxGroupIndex,
+        },
+
+        ...entries,
+
+        // Last tick
+        {
+          formattedValue: undefined,
+          offset: scale.range()[1],
+          labelOffset: 0,
+          groupIndex: maxGroupIndex,
+        },
+      ];
+    }
+
+    // scale type = 'point'
+    const { entries } = mapToGrouping(
+      filteredDomain,
+      getGrouping,
+      tickPlacement ?? 'extremities',
+      tickLabelPlacement ?? 'middle',
+      scale,
+    );
+    return entries;
+  }, [scale, tickInterval, getGrouping, tickPlacement, tickLabelPlacement]);
+}
+
+function mapToGrouping(
+  tickValues: any[],
+  getGrouping: (value: any, dataIndex: number) => any[],
+  tickPlacement: Exclude<TickParams['tickPlacement'], undefined>,
+  tickLabelPlacement: Exclude<TickParams['tickLabelPlacement'], undefined>,
+  scale: D3Scale,
+) {
+  // Step 1: Create all tick items with their group information
+  const allTickItems: Omit<Required<GroupedTickItemType>, 'syncIndex'>[] = [];
+
+  let maxGroupIndex = 0;
+
+  // Build all tick items from the input data
+  tickValues.forEach((tickValue, dataIndex) => {
+    const groupValues = getGrouping(tickValue, dataIndex);
+
+    groupValues.forEach((groupValue, groupIndex) => {
+      maxGroupIndex = Math.max(maxGroupIndex, groupIndex);
+
+      allTickItems.push({
+        value: groupValue,
+        formattedValue: `${groupValue}`,
+        offset: isBandScale(scale)
+          ? scale(tickValue)! -
+            (scale.step() - scale.bandwidth()) / 2 +
+            offsetRatio[tickPlacement] * scale.step()
+          : scale(tickValue),
+        groupIndex,
+        dataIndex,
+        ignoreTick: false,
+        labelOffset: 0,
+      });
+    });
+  });
+
+  // Step 2: Sort by group index to ensure proper ordering
+  allTickItems.sort((a, b) => a.groupIndex - b.groupIndex);
```

Comment:
You mapped the new `groups` structure to generate `getGrouping` and keep this internal logic. BU tit might be simpler to directly use `groups`

```js
const maxGroupIndex = groups.length -1

const allTickItems = []

for (let groupIndex = 0; groupIndex < groups.length; groupIndex+=1){
  // No need to sort anymore

  tickValues.forEach((tickValue, dataIndex) => {
    const groupValue = groups[groupIndex].getValue(tickValue, dataIndex);

    // ....

    if(isNew){
      // Add a new item
      allTickItems.push( ... )
    } else {
      // Merge it with the last item
      allTickItems[allTickItems.length -1].size += itemSize
    }
  })
}
```

Reply:
https://github.com/mui/mui-x/issues/19067